From: Yu Zhao <yuzhao@google.com>
Subject: mm/vmscan: use the proper VMA iterator
Date: Mon, 12 Sep 2022 14:01:28 -0600

When get_next_vma() finishes iterating VMAs within a range [start, end),
it expects args->vma to point the first VMA out of that range, if such a
VMA exists.  This allows its callers to continue the iteration with a new
range above the previous one, if those callers choose to.

for_each_vma_range() always sets args->vma to NULL after it's done.  This
may mislead those callers to conclude that there are no more VMAs, and in
turn they terminate their iterations prematurely.

This fix replaces for_each_vma_range() with for_each_vma() and explicitly
checks whether the next VMA is still within range, and if not, returns
false to indicate the current range has ended.  The callers may continue
with the next range if args->vma is not NULL.

Link: https://lkml.kernel.org/r/Yx+QGOgHg1Wk8tGK@google.com
Signed-off-by: Yu Zhao <yuzhao@google.com>
Cc: Liam Howlett <liam.howlett@oracle.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/vmscan.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

--- a/mm/vmscan.c~mm-vmscan-use-vma-iterator-instead-of-vm_next-fix
+++ a/mm/vmscan.c
@@ -3781,7 +3781,10 @@ static bool get_next_vma(unsigned long m
 	VM_WARN_ON_ONCE(mask & size);
 	VM_WARN_ON_ONCE((start & mask) != (*vm_start & mask));
 
-	for_each_vma_range(vmi, args->vma, end) {
+	for_each_vma(vmi, args->vma) {
+		if (end && end <= args->vma->vm_start)
+			return false;
+
 		if (should_skip_vma(args->vma->vm_start, args->vma->vm_end, args))
 			continue;
 
_
